### PR TITLE
Added Mainnet Contract

### PR DESCRIPTION
**Pull request by span14**

### CODE REVIEW

**Review by wgr523 — COMMENTED** [review]
LGTM

**Review by liam-lai — APPROVED** [review]

File: mainnet_contract/contracts/HeaderReader.sol [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+    uint256 memPtr;
+  }
+
+  function getParentHashAndNumber(bytes memory genesis_header) public pure returns (bytes32, int) {
+    RLPItem[] memory ls = toList(toRlpItem(genesis_header));
+    return (toBytes32(toBytes(ls[0])), int(toUint(ls[8])));
+  }
+
+  // Returns (parentHash, number, round_number, certificate hash, sigs)
+  function getValidationParams(bytes memory header) public pure returns (bytes32, int, uint64, bytes32, bytes[] memory) {
```
Comment (wgr523) [inline mainnet_contract/contracts/HeaderReader.sol]: use int and uint64, why? (why not all uint256?)

File: mainnet_contract/contracts/HeaderReader.sol [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+  function toBytes32(bytes memory data) internal pure returns (bytes32 res) {
+    assembly {
+      res := mload(add(data, 32))
+    }
+  }
+
+  /*
+  * @param item RLP encoded bytes
+  */
+  function toRlpItem(bytes memory item) internal pure returns (RLPItem memory) {
```
Comment (wgr523) [inline mainnet_contract/contracts/HeaderReader.sol:82]: Starting from here, are these functions from some other source? If so, please give the reference. Then we can check whether the origin could have a bug or keep the code up to date.

File: mainnet_contract/contracts/Subnet.sol [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+  event SubnetBlockFinalized(bytes32 block_hash, int number);
+
+  // Modifier
+  modifier onlyMasters() {
+    if (!masters[msg.sender]) revert("Masters Only");
+    _;
+  }
+
+  constructor(address[] memory initial_validator_set, int threshold, bytes memory genesis_header) public {
+    require(initial_validator_set.length > 0, "Validator Empty");
```
Comment (wgr523) [inline mainnet_contract/contracts/Subnet.sol:49]: require(threshold > 0, "Invalid Threshold");

File: mainnet_contract/contracts/Subnet.sol [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    validator_sets[subnet_block_height] = Validators({
+      set: new_validator_set,
+      threshold: threshold
+    });
+  }
+
+  function receiveHeader(bytes memory header) public onlyMasters { 
+    (bytes32 parent_hash, int number, uint64 round_number, bytes32 signHash, bytes[] memory sigs) = HeaderReader.getValidationParams(header);
+    require(number > 0, "Repeated Genesis");
+    require(header_tree[parent_hash].hash != 0, "Parent Missing");
```
Comment (wgr523) [inline mainnet_contract/contracts/Subnet.sol:92]: Is default hash 0? Or is it 0000....0000?

File: mainnet_contract/contracts/Subnet.sol [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+      }
+      current_validator_set_pointer = number;
+    }
+
+    int unique_counter = 0;
+    address[] memory signer_list = new address[](sigs.length);
+    for (uint i = 0; i < sigs.length; i++) {
+      address signer = recoverSigner(signHash, sigs[i]);
+      if (lookup[signer] != true) {
+        continue;
```
Comment (wgr523) [inline mainnet_contract/contracts/Subnet.sol]: We can enforce the header contain no non-validator signatures, so here we can revert.

File: mainnet_contract/contracts/Subnet.sol [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+      }
+      if (!unique_addr[signer]) {
```
Comment (wgr523) [inline mainnet_contract/contracts/Subnet.sol:141]: We can enforce the header contain no repeated signatures, so here we can add a `else { revert(...) }`. And the logic here will also be simpler

File: mainnet_contract/contracts/Subnet.sol [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    require(sig.length == 65, "Invalid Signature");
+    assembly {
+      // first 32 bytes, after the length prefix.
+      r := mload(add(sig, 32))
+      // second 32 bytes.
+      s := mload(add(sig, 64))
+      // final byte (first byte of the next 32 bytes).
+      v := byte(0, mload(add(sig, 96)))
+    }
+    // TOCHECK: v needs 27 more, may related with EIP1559
```
Comment (wgr523) [inline mainnet_contract/contracts/Subnet.sol:198]: I remember 27 is a Ethereum defined number.

File: mainnet_contract/test/subnet.js [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+      RLP.encode([
+        ["0x0000000000000000000000000000000000000000000000000000000000000000", 0, 0],
+        0
+      ])
+    ));
+    const raw_sigs = [];
+    for (let i = 0; i < 3; i++) {
+      raw_sigs.push(
+        secp256k1.ecdsaSign(
+          hex2Arr(voteForSignHash.substring(2)),
```
Comment (wgr523) [inline mainnet_contract/test/subnet.js:138]: what's the meaning of substring(2)

File: mainnet_contract/test/subnet.js [hunk 9]
```diff
@@ -0,0 +1,600 @@
...
+          hex2Arr(this.validators[i].privateKey.substring(2))
+        )
+      );
+    }
+    const sigs = raw_sigs.map(x => {
```
Comment (wgr523) [inline mainnet_contract/test/subnet.js:143]: what's the meaning of this step

File: mainnet_contract/contracts/Subnet.sol [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+    bytes32 parent_hash;
+    bool finalized;
+    uint mainnet_num;
+    bytes src;
+  }
+
+  struct Validators {
+    address[] set;
+    int threshold;
+  }
```
Comment (wgr523) [inline mainnet_contract/contracts/Subnet.sol:22]: two separate setting method. address[] is from new Header.Validators.NextEpoch. threshold is from (grand)master, so there should be a new func SetThreshold(at int, threhold int)

File: mainnet_contract/contracts/Subnet.sol [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    for (uint i = 0; i < 3; i++) {
+      if (header_tree[curr_hash].parent_hash == 0) return;
+      bytes32 prev_hash = header_tree[curr_hash].parent_hash;
+      if (header_tree[curr_hash].round_num != header_tree[prev_hash].round_num+1) return;
+      curr_hash = prev_hash;
+    }
+    latest_finalized_block = curr_hash;
+    // Confirm all ancestor unconfirmed block
+    while (header_tree[curr_hash].finalized != true) {
+      header_tree[curr_hash].finalized = true;
```
Comment (wgr523) [inline mainnet_contract/contracts/Subnet.sol:177]: here, the block header.Validators is set to nextEpochValidators

File: mainnet_contract/contracts/Subnet.sol [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+    require(threshold > 0, "Invalid Threshold");
+    require(subnet_block_height >= current_validator_set_pointer, "Modify Past Validator");
+    validator_sets[subnet_block_height] = Validators({
+      set: new_validator_set,
+      threshold: threshold
+    });
+  }
+
+  function receiveHeader(bytes memory header) public onlyMasters { 
+    (bytes32 parent_hash, int number, uint64 round_number, bytes32 signHash, bytes[] memory sigs) = HeaderReader.getValidationParams(header);
```
Comment (wgr523) [inline mainnet_contract/contracts/Subnet.sol]: read two more: currentEpochValidators, next...

- currentEpochValidators = nil for all !IsEpochSwitch(block).
- nextEpochValidators = nil for all block: number % EPOCH != EPOCH-GAP(+1).

Need to check: when encounter currentEpochValidators, check == last committed nextEpochValidators

File: mainnet_contract/contracts/Subnet.sol [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+    require(new_validator_set.length > 0, "Validator Empty");
+    require(threshold > 0, "Invalid Threshold");
+    require(subnet_block_height >= current_validator_set_pointer, "Modify Past Validator");
+    validator_sets[subnet_block_height] = Validators({
+      set: new_validator_set,
+      threshold: threshold
+    });
+  }
+
+  function receiveHeader(bytes memory header) public onlyMasters {
```
Comment (wgr523) [inline mainnet_contract/contracts/Subnet.sol:107]: do we really need modifier `onlyMasters`?

**Comment by liam-lai** [comment]
no need

**Pull request closed without merging**